durango/src/durango.rs

```diff
@@ -42,7 +43,10 @@ impl Durango {
     /// Read all the bytes in the file.
     /// Return Ok(vec) if succ
     /// Otherwise return Err(msg) with the error message as String
-    fn read_all_bytes_in_file(filename: &str) -> Result<Vec<u8>, DurangoError> {
+    fn read_all_bytes_in_file<P>(filename: P) -> Result<Vec<u8>, DurangoError>
+    where
+        P: AsRef<path::Path>
```

Comment:
Is the `where P: AsRef<path::Path>` strictly necessary here?

If not, it adds unnecessary cognitive load for the reader.

Reply:
There is an alternative:
``` rust
read_all_bytes_in_file<P: AsRef<path::Path>>(filename: P) -> blablabla
```

Which while shorter, but the nested angle brackets can get out of hand. The `where` clause does a good job of avoiding that IMO.

Reply:
Unfortunately you can't get rid of the declared "P", since AsRef is a trait and not a type. The tradeoff of extra noise here is less noise in the user's code.

Reply:
Does `read_all_bytes_in_file(filename: impl AsRef<path::Path>) -> blablabla` ?
A quote from [this link](https://doc.rust-lang.org/edition-guide/rust-2018/trait-system/impl-trait-for-returning-complex-types-with-ease.html)

Reply:
It doesn't! Well it does but impl args have problems...

`impl` arguments have this glaring issue that they change the type of the function.

So `read_all_bytes_in_file(filename: impl AsRef<path::Path>)` and `read_all_bytes_in_file<P: AsRef<path::Path>>(filename: P)` actually have different types, making it a backwards compatibility issue to change between the two.

A specific example where this causes problems is that developers often expect to be able to explicitly specify the type in template functions in order to fix type-inference issues:

``` rust
// explicit type to avoid "ambiguous type"
read_all_byte_in_file::<OsPath>(parse_from_message());
```

This doesn't work with impl args.

The advice I've read is to only use impl for return statements, where it does clean up code significantly. So I've adopted that practice for most of my code.

--- 

But I'm happy to switch to `impl AsRef<path::Path>` if we want, it does simplify the code here. But it comes with some frustrating baggage.

Reply:
I see. It makes sense to not use `impl` in parameters. I don't mind both ways, as long as it lines up with your command-line change.